compiler/rustc_codegen_ssa/src/back/link.rs

```diff
@@ -3116,12 +3116,31 @@ fn add_lld_args(
 
     let self_contained_linker = self_contained_cli || self_contained_target;
     if self_contained_linker && !sess.opts.cg.link_self_contained.is_linker_disabled() {
-        for path in sess.get_tools_search_paths(false) {
-            cmd.arg({
-                let mut arg = OsString::from("-B");
-                arg.push(path.join("gcc-ld"));
-                arg
-            });
+        // We tell `cc` where to find the lld wrappers, in a similar way to how we locate
+        // codegen-backends dylibs:
+        // - in the tool search paths: the sysroot's `rustlib` bin path,
+        // - and as a fallback for cases where `--sysroot` only contains a target std: in the
+        //   default host sysroot where rustc is currently located.
+        let fallback_sysroot_paths = filesearch::sysroot_candidates()
+            .into_iter()
+            .map(|sysroot| filesearch::make_target_bin_path(&sysroot, config::host_triple()));
+
+        let mut linker_path_exists = false;
+        for path in sess.get_tools_search_paths(false).into_iter().chain(fallback_sysroot_paths) {
+            let linker_path = path.join("gcc-ld");
+            if linker_path.exists() {
+                linker_path_exists = true;
+                cmd.arg({
+                    let mut arg = OsString::from("-B");
+                    arg.push(linker_path);
+                    arg
+                });
+            }
+        }
+        if !linker_path_exists {
+            // As a sanity check, we emit an error if none of these paths exist: we want
+            // self-contained linking and have no linker.
+            sess.dcx().emit_fatal(errors::SelfContainedLinkerMissing);
```

Comment:
What if a distro wants to drop Rust lld binaries and rely on system lld?

Then the corresponding `self-contained` directories either won't exist or will be empty.
If we just push the `-B` options without checking then gcc will go through them and then proceed searching and finding lld on the system without errors.

If we report an error here, then such distros will also need to reset self-contained flags in target specs to false, probably through env vars like the current `CFG_USE_SELF_CONTAINED_LINKER`.
This will have a negative effect on portability of options like `-Clink-self-contained=+linker` between distros (if we need such portability).

Reply:
But I'm fine with reporting a conservative error here for now, it can always be relaxed later.
At least we'll be able see what breaks.

Reply:
Another reason why I preferred the fallback compared to the error :3, because falling back to system lld is what we currently do in the case you mention of course.

But one can also say having the self-contained linker enabled is an opt-in for the toolchain maker, it goes hand in hand with bootstrap building lld and putting in the sysroot. Not having both can be seen as a mismatch, and that situation would be properly modeled as `-Clink-self-contained=-linker -Clinker-features=+lld`.

Reply:
Note that I only check for the existence of the `gcc-ld` folder, so in theory an empty directory with no lld binaries would pass this test 🤡

Reply:
One problem with the fallback is that if there is no system lld and we try to use it anyway, the error is *really bad*:
```
  = note: collect2: fatal error: cannot find 'ld'
          compilation terminated.
```
It doesn't even say that it was trying to find `lld`, mentioning `ld` instead which is not what it was looking for.

Reply:
> What if a distro wants to drop Rust lld binaries and rely on system lld?

Sounds to me like they should then disable self-contained linking in their rustc builds.

Reply:
It's definitely not going to be an oscar of the best error message, but also it's in gcc, people are used to it, and look for the `-fuse-ld` option that is in the command just above that note.

Reply:
> a non-zero number of people are used to it, and look for the -fuse-ld option that is in the command just above that note. (Maybe it also depends on the gcc version and newer ones have improved it 🤔)

I dare say most Rust users are not used to it. I certainly was entirely flabbergasted, and I know more than the average Rust programmer about how the compiler works. If an error is too obscure for me I think we can safely assume that it is useless-or-worse for the majority of our users.

That should obviously be reported against gcc, I just didn't have the time for that. But in general we try to work around weaknesses of other tools we employ when we are aware of them.

But as you say indeed for this PR this isn't relevant. I was just replying to @petrochenkov's arguments.